pkg/reconciler/stats_reporter.go

```diff
@@ -0,0 +1,172 @@
+	// View name defaults to the measure name if unspecified.
+	if err := view.Register(
+		&view.View{
+			Description: InstallSuccessCountStat.Description(),
+			Measure:     InstallSuccessCountStat,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: InstallSuccessLatencyStat.Description(),
+			Measure:     InstallSuccessLatencyStat,
+			Aggregation: view.Distribution(10, 100, 1000, 10000, 30000, 60000), // Bucket boundaries are 10ms, 100ms, 1s, 10s, 30s and 60s.
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: DeploymentReadyCountStat.Description(),
+			Measure:     DeploymentReadyCountStat,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: DeploymentReadyLatencyStat.Description(),
+			Measure:     DeploymentReadyLatencyStat,
+			Aggregation: view.Distribution(10, 100, 1000, 10000, 30000, 60000), // Bucket boundaries are 10ms, 100ms, 1s, 10s, 30s and 60s.
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+	); err != nil {
+		panic(err)
+	}
+}
+
+// StatsReporter defines the interface for sending metrics
+type StatsReporter interface {
+	// ReportKnativeServingReady reports the count and latency metrics for a reconcile operation
+	ReportInstallSuccess(resourceNamespace, resourceName string, duration time.Duration) error
+	ReportDeploymentReady(resourceNamespace, resourceName string, duration time.Duration) error
+}
+
+// Reporter holds cached metric objects to report metrics
+type reporter struct {
+	reconciler string
+	ctx        context.Context
+}
+
+// srKey is used to associate StatsReporters with contexts.
+type srKey struct{}
+
+// WithStatsReporter attaches the given StatsReporter to the provided context
+// in the returned context.
+func WithStatsReporter(ctx context.Context, sr StatsReporter) context.Context {
+	return context.WithValue(ctx, srKey{}, sr)
+}
+
+// GetStatsReporter attempts to look up the StatsReporter on a given context.
+// It may return null if none is found.
+func GetStatsReporter(ctx context.Context) StatsReporter {
+	untyped := ctx.Value(srKey{})
+	if untyped == nil {
+		return nil
+	}
+	return untyped.(StatsReporter)
+}
+
+// NewStatsReporter creates a reporter that collects and reports metrics
+func NewStatsReporter(reconciler string) (StatsReporter, error) {
+	// Reconciler tag is static. Create a context containing that and cache it.
+	ctx, err := tag.New(
+		context.Background(),
+		tag.Insert(reconcilerTagKey, reconciler))
+	if err != nil {
+		return nil, err
+	}
+
+	return &reporter{reconciler: reconciler, ctx: ctx}, nil
+}
+
+// ReportInstallSuccess reports
+func (r *reporter) ReportInstallSuccess(resourceNamespace, resourceName string, duration time.Duration) error {
+	key := fmt.Sprintf("%s/%s", resourceNamespace, resourceName)
```

Comment:
generating the context should be a separate function that is shared with other metrics calls.

Reply:
Refactored